File: lab1a/cpp/limited_buffer.cpp
#include <vector>
#include <mutex>
#include <condition_variable>

#include "common.h"

class LimitedBuffer {
	std::vector<DataEntry> items;
	int capacity;
	std::mutex items_mutex;
	std::condition_variable items_cv;

	int reserved_count;
public:
	int estimated_count;

	LimitedBuffer(int capacity) {
		this->capacity = capacity;
		estimated_count = 0;
		reserved_count = 0;
	}

	void insert(DataEntry entry);
	DataEntry remove();

	bool reserve_entry();
};

void LimitedBuffer::insert(DataEntry entry) {
	std::unique_lock<std::mutex> guard(items_mutex);
	items_cv.wait(guard, [&](){ return items.size() < this->capacity; });

	items.push_back(entry);
	items_cv.notify_all();
}

DataEntry LimitedBuffer::remove() {
	std::unique_lock<std::mutex> guard(items_mutex);
	items_cv.wait(guard, [&](){ return items.size() > 0; });

	DataEntry entry = items[items.size()-1];
	items.pop_back();
	items_cv.notify_all();
	return entry;
}

bool LimitedBuffer::reserve_entry() {
	std::unique_lock<std::mutex> guard(items_mutex);

	if (reserved_count == estimated_count) {
		return false;
	} else {
		reserved_count++;
		return true;
	}
}

File: lab1a/go/limited_buffer.go
package main

import "sync"

type LimitedBuffer struct {
	items       []DataEntry
	readIndex   int
	writeIndex  int
	currentSize int
	mutex       *sync.Mutex
	updateCond  *sync.Cond

	estimatedCount int
	reservedCount int
}

func makeLimitedBuffer(bufferSize int) LimitedBuffer {
	var container = make([]DataEntry, bufferSize)

	var itemsMutex = sync.Mutex{}
	var cond = sync.NewCond(&itemsMutex)
	return LimitedBuffer{
		items: container,
		mutex: &itemsMutex,
		updateCond: cond,
	}
}

func (buffer *LimitedBuffer) Insert(item DataEntry) {
	buffer.mutex.Lock()
	for buffer.currentSize == len(buffer.items) {
		buffer.updateCond.Wait()
	}
	buffer.items[buffer.writeIndex] = item
	buffer.writeIndex = (buffer.writeIndex + 1) % len(buffer.items)
	buffer.currentSize++
	buffer.updateCond.Broadcast()
	buffer.mutex.Unlock()
}

func (buffer *LimitedBuffer) Remove() DataEntry {
	buffer.mutex.Lock()
	for buffer.currentSize == 0 {
		buffer.updateCond.Wait()
	}
	var item = buffer.items[buffer.readIndex]
	buffer.readIndex = (buffer.readIndex + 1) % len(buffer.items)
	buffer.currentSize--
	buffer.updateCond.Broadcast()
	buffer.mutex.Unlock()
	return item
}

func (buffer *LimitedBuffer) ReserveEntry() bool {
	buffer.mutex.Lock()
	defer buffer.mutex.Unlock()

	if buffer.estimatedCount == buffer.reservedCount {
		return false
	} else {
		buffer.reservedCount++
		return true
	}
}
